app/upload.go

```diff
@@ -19,6 +19,67 @@ import (
 
 const minFirstPartSize = 5 * 1024 * 1024 // 5MB
 
+func (a *App) runPluginsHook(info *model.FileInfo, file io.Reader) *model.AppError {
+	pluginsEnvironment := a.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return nil
+	}
+
+	// using a pipe to avoid loading the whole file content in memory.
+	r, w := io.Pipe()
+	errChan := make(chan *model.AppError, 1)
+	go func() {
+		defer w.Close()
+		defer close(errChan)
+		pluginContext := a.PluginContext()
+		pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+			newInfo, rejStr := hooks.FileWillBeUploaded(pluginContext, info, file, w)
+			if rejStr != "" {
+				errChan <- model.NewAppError("runPluginsHook", "app.upload.run_plugins_hook.rejected", nil, rejStr, http.StatusBadRequest)
+				return false
+			}
+			if newInfo != nil {
+				info = newInfo
```

Comment:
Isn't there a race here? We are writing to info from a goroutine, but reading `info.Path` at L48.

Reply:
It looks like it. We should make a copy of the path before launching the goroutine. Great catch :+1: